File: server/StructFlow/Sorts/QuickSort.py
def quick_sort(array):
    steps = []
    arr = array.copy()

    def record_step(highlight_indices, swapped=False, comparison=None, values=None):
        steps.append({
            "array": arr.copy(),
            "highlight": highlight_indices,
            "swapped": swapped,
            "comparison": comparison,
            "values": values  # הערכים שהושוו בפועל
        })

    def median_of_three(low, high):
        mid = (low + high) // 2
        trio = [(arr[low], low), (arr[mid], mid), (arr[high], high)]
        trio.sort()
        return trio[1][1]

    def partition(low, high):
        pivot_index = median_of_three(low, high)
        pivot = arr[pivot_index]

        arr[pivot_index], arr[high] = arr[high], arr[pivot_index]
        record_step([pivot_index, high], swapped=True,
                    comparison=f"pivot swap: {arr[high]} ↔ {arr[pivot_index]}",
                    values=[arr[high], arr[pivot_index]])

        i = low - 1
        for j in range(low, high):
            if arr[j] < pivot:
                i += 1
                if i != j:
                    arr[i], arr[j] = arr[j], arr[i]
                    record_step([i, j], swapped=True,
                                comparison=f"{arr[j]} < pivot({pivot}) → True → swap {arr[i]} ↔ {arr[j]}",
                                values=[arr[i], arr[j]])
                else:
                    record_step([i, j], swapped=False,
                                comparison=f"{arr[j]} < pivot({pivot}) → True → already in position",
                                values=[arr[j], arr[j]])
            else:
                record_step([j, high], swapped=False,
                            comparison=f"{arr[j]} < pivot({pivot}) → False → no swap",
                        values=[arr[j], pivot])


        if i + 1 == high:
            record_step([i + 1, high], swapped=False,
                        comparison=f"pivot {arr[high]} is already in place",
                        values=[arr[high], arr[high]])
        else:
            arr[i + 1], arr[high] = arr[high], arr[i + 1]
            record_step([i + 1, high], swapped=True,
                        comparison=f"Moved pivot {pivot} to position {i + 1}",
                        values=[arr[i + 1], arr[high]])

        return i + 1

    def quick_sort_rec(low, high):
        if low < high:
            pi = partition(low, high)
            quick_sort_rec(low, pi - 1)
            quick_sort_rec(pi + 1, high)

    quick_sort_rec(0, len(arr) - 1)
    return {
        "steps": steps,
        "sorted_array": arr
    }

File: client/src/Sorts/QuickSort.js
import { gsap } from "gsap";

export const runQuickSort = async ({
  arrayData,
  setArrayData,
  setHighlightIndices,
  setSwappedIndices,
  setLogs,
  setSortActive,
  offsetsRef,
  forceRender,
  ARRAY_API,
}) => {
  setSortActive(true);

  const result = await ARRAY_API.QuickSort();
  const steps = result.steps;
  const finalSortedArray = result.sorted_array;

  let i = 0;
  let localArray = [...arrayData];

  const processStep = () => {
    if (i >= steps.length) {
      setHighlightIndices([]);
      setSwappedIndices([]);
      setArrayData([...finalSortedArray]);
      setSortActive(false);
      return;
    }

    const step = steps[i];
    const [i1, i2] = step.highlight;
    const [val1, val2] = step.values || [localArray[i1], localArray[i2]];
    const { comparison, swapped } = step;

    setHighlightIndices(step.highlight);
    setSwappedIndices(swapped ? step.highlight : []);

    const key1 = `${i1}`;
    const key2 = `${i2}`;
    offsetsRef.current[key1] = 0;
    offsetsRef.current[key2] = 0;

   const logEntry = (
    <span key={i} style={{ lineHeight: "1.6" }}>
        {comparison && (
        <>
            <strong>Pivot:</strong> {val2} (index {i2})<br />
            <strong>Comparing:</strong> {val1} (index {i1}) with pivot<br />
        </>
        )}

        {comparison?.includes("True") && !swapped ? (
        <>
            Result: {val1} is smaller than the pivot, but it's already in the correct place. No swap needed.
        </>
        ) : comparison?.includes("True") && swapped ? (
        <>
            Result: {val1} is smaller than the pivot. Swapped positions {i1} and {i2}.
        </>
        ) : comparison?.includes("False") ? (
        <>
            Result: {val1} is not smaller than the pivot. No swap performed.
        </>
        ) : swapped ? (
        <>
            Swapped values at indices {i1} and {i2} to place the pivot in its correct position.
        </>
        ) : (
        <>
            No swap was needed.
        </>
        )}
    </span>
    );


    setLogs(prev => [...prev, logEntry]);

    if (swapped) {
      const dx = (i2 - i1) * 80;

      gsap.to(offsetsRef.current, {
        [key1]: dx,
        [key2]: -dx,
        duration: 1.5,
        ease: "power2.inOut",
        onUpdate: forceRender,
        onComplete: () => {
          [localArray[i1], localArray[i2]] = [localArray[i2], localArray[i1]];
          setArrayData([...localArray]);
          offsetsRef.current[key1] = 0;
          offsetsRef.current[key2] = 0;
          forceRender();
          i++;
          setTimeout(processStep, 1000);
        }
      });
    } else {
      i++;
      setTimeout(processStep, 1500);
    }
  };

  processStep();
};
